### ISSUE TITLE
[SW SQL Server Products] - Show License Keys

### ISSUE THREAD

**Issue by Ioan-Popovici** [MEM-Zone/sccm-zone.com#5]
Hi Ioan,
 
Firstly I would like to say Thank you very much for writing the ‘SQL Version Detection and Reporting with SCCM’ article.  We have implemented the code and are currently monitoring the results.
 
I have been asked by my manager to “get all the SQL Licences/product keys for every SQL instance”.  Would it be possible for you to modify the code so that the SQL Licences/product keys are pulled from the registry and parsed using your SCCM code?
 
JAKOB BINDSLET has posted some PS code which parses the key from the registry:
 
http://mspowershell.blogspot.com/2010/11/sql-server-product-key.html
 
This would be extremely helpful to me (and many others I suspect).
 
Again – Many Thanks,
Matthew Lue

**Comment by Ioan-Popovici** [comment]
I have found a way to do it but that does requires another hardware extension. You will need to enable gathering of a few registry keys. The problem is that the key is obfuscated and will require a sql or report function. Will probably figure it out by the end of next week.

**Comment by marshgm** [comment]
Have you considered in the reports for SQL version, to also add what Windows OS SQL is installed on?

**Comment by matthewgnslue** [comment]
> Have you considered in the reports for SQL version, to also add what Windows OS SQL is installed on?

This can already be attained by joining using [ResourceID] to [v_R_System].[Operating_System_Name_and0] right?

**Comment by Ioan-Popovici** [comment]
>Have you considered in the reports for SQL version, to also add what Windows OS SQL is installed on?

@marshgm Yes I will completely rewrite that report and possibly the detection method.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Ioan-Popovici** [comment]
>> Have you considered in the reports for SQL version, to also add what Windows OS SQL is installed on? ...

@matthewgnslue It’s a bit more complicated if you want to make sure you get the caption especially when having hwi issues. I already have a method for this.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Ioan-Popovici** [comment]
Had some work related tasks that took priority over the blog requests.
These are now complete, resuming work...

**Comment by Ioan-Popovici** [comment]
Simplified WMI WQL query, selecting only what’s needed, much less noise now.
Need to add SRS, IP, Port and license info.
Will try to make one WMI instance per SQL instance but I don’t know if it’s possible with WQL.
If we cant have everything neatly in one place, I will move the processing in the reporting query.

**Comment by Ioan-Popovici** [comment]
Dropped SRS, IP, Port for now. I don't see the value of adding these. Feel free to comment if you disagree.
Optimized WQL queries and definitions. Updated HWI classes so that they do not clash with the old ones.
Remaining task on the extension side:
* Add License info
* Better naming for database tables and views

**Comment by Ioan-Popovici** [comment]
Just committed the final version of the WMI extension
I’m pretty sure that ProductID info will not work for anything below SQL 2012

Remaining tasks:
* SQL Query
* SSRS Report

**Comment by Ioan-Popovici** [comment]
This was a bit complicated due to my limited understanding of how SQL does stuff.

In order to simplify the query and not duplicate code, I created a [stored procedure](https://github.com/SCCM-Zone/sccm-zone.github.io/blob/master/Reporting/User_Defined_Functions/usp_PivotWithDynamicColumns.sql) for pivoting SQL WMI data directly that supports dynamic columns.

Before:
![Screen Shot 2019-06-14 at 20 44 55](https://user-images.githubusercontent.com/17069425/59527833-5a96d480-8ee5-11e9-852f-20f5439415f7.png)

After:
![Screen Shot 2019-06-14 at 20 43 58](https://user-images.githubusercontent.com/17069425/59527863-6a161d80-8ee5-11e9-880f-e9f2cbacb51c.png)

## Remaining tasks:

* SQL Query    - Should be much easier now     
* SSRS Report - Template created

**Comment by Ioan-Popovici** [comment]
Released v2.0-prerelease of the SQL Products query. Report will follow.

# This is a pre-release version!

Completley re-written, adds functionality to see Product Key and many other SQL properties. I will add a changelog later.

## Known issues:

 * Version0 and Version do not store the same value type in all tables. Need to modify the mof extension.

## Remaining tasks:

* mof file bugfix
* SSRS Report
* SSRS Report script to de-obfuscate the product key. I kept it obfuscated in the DB but this report should be restricted for obvious reasons.

Query result:
![Screen Shot 2019-06-18 at 20 00 03](https://user-images.githubusercontent.com/17069425/59705159-7f04f080-9206-11e9-8a51-c840a0344eb8.png)

**Comment by Ioan-Popovici** [comment]
#Added new fields
Domain, OS, IsVirtualMachine, CPUs, CPUCores, LogicalProcessors
Have a look and let me know if anything else is needed:
![Screen Shot 2019-06-19 at 18 43 04](https://user-images.githubusercontent.com/17069425/59780711-16c81480-92c3-11e9-90fc-3657c559c2ba.png)

**Comment by matthewgnslue** [comment]
Looks great to me - it's got all the information I need :-)

**Comment by marshgm** [comment]
Looks good. This has everything I need, this will be a big help. 
Thank you...

**Comment by Ioan-Popovici** [comment]
Ok, now I have to figure out how to get that stored procedure in the report, it seems it does not work with the inline query in Report builder, although it works fine in SSMS. If I can’t figure it out (unlikely) you might need to add that stored procedure manually (paste in SSMS + Run). I hope that the report will be ready tomorrow so you guys will be the fist to test it until I update the blog post :D

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Ioan-Popovici** [comment]
Hi guys , I'm about to finalize the design, can you please check if the report has everything you need?
I removed some of the fields that I thought not to be useful. This is your chance to change my mind :)
# Report

![Screen Shot 2019-06-24 at 18 21 45](https://user-images.githubusercontent.com/17069425/60031785-05f61500-96ae-11e9-858b-3931fbfdd60b.png)

# All fields

![59705159-7f04f080-9206-11e9-8a51-c840a0344eb8](https://user-images.githubusercontent.com/17069425/60031811-11e1d700-96ae-11e9-8b69-0dcd05bae6bc.png)

Thanks
@marshgm @matthewgnslue @matt2005

PS: The key de-obfuscation code is not written yet, I have to translate it from powershell that's why you see the key as a bit string value.

**Comment by matthewgnslue** [comment]
Hi @Ioan-Popovici .  The only item no longer present which I think may be important to licensing is "Clustered" state.  Other than that - all good for me.

**Comment by marshgm** [comment]
It looks good.

**Comment by Ioan-Popovici** [comment]
Will add the clustered back, thanks for the feedback :)

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Ioan-Popovici** [comment]
Hi @matthewgnslue,

	I finished the report but I’m having trouble with the VB.net script for the key deobfuscation.
I’ve tested it in an online compiler and it works but in the report it keeps returning  #Error.
I don’t have visual studio so I can do proper debugging but I will install it tomorrow because like this I have zero error output.

Meanwhile please have a look at it, I've been at it all day. My VB.net knowledge is rusty at best.

# VB Function

```vbnet
Function DeobfuscateSQLKey(ByVal sqlKeyBinary As Byte(), ByVal sqlVersion As Integer) As String
    '/* Test variables. !! Must be comented for production'
    'sqlKeyBinary = {1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0}
    'sqlVersion = 12
    Dim charsArray As Char() = {"B","C","D","F","G","H","J","K","M","P","Q","R","T","V","W","X","Y","2","3","4","6","7","8","9"}
    Dim sqlProductKey As String
    Dim sqlKeyBinaryIndex As Integer
    Dim i, j, k As Integer
    Try
        If (sqlVersion >= 11) Then
            sqlKeyBinaryIndex = 0
        Else
            sqlKeyBinaryIndex = 52
        End If
        For i = 24 To 0 Step -1
            k = 0
            For j = 14 To 0 Step -1
                k = k * 256 Xor sqlKeyBinary(j + sqlKeyBinaryIndex)
                sqlKeyBinary(j + sqlKeyBinaryIndex) =  Math.Truncate(k / 24)
                k = k Mod 24
            Next j
            sqlProductKey = charsArray(k) + sqlProductKey
            If (i Mod 5) = 0 And i <> 0 Then
                sqlProductKey = "-" + sqlProductKey
            End If
        Next i
    Catch
        sqlProductKey = "Cannot decode product key."
    End Try
        DeobfuscateSQLKey = sqlProductKey
End Function
```

# Report Call

To call it in any report you need something like this:
```
=Code.DeobfuscateSQLKey(Fields!ProductKey.Value,12)
```

# Online Compiler Result

https://onlinegdb.com/SkWx2R1eS

# Report Preview 

The final report looks something like this:

![Screen Shot 2019-06-25 at 20 24 59](https://user-images.githubusercontent.com/17069425/60119653-a2401a80-9787-11e9-8cc6-a15989dad624.png)

**Comment by matthewgnslue** [comment]
Sorry @Ioan-Popovici .  I don't have any VB knowledge, so I can't help on this unfortunately.

**Comment by Ioan-Popovici** [comment]
>Sorry @Ioan-Popovici .  I don't have any VB knowledge, so I can't help on this unfortunately.

@matthewgnslue Eh... ‘tis but a flesh wound. I’ll figure it out once I can have some error reporting with vb studio :)

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Ioan-Popovici** [comment]
Report is finished, will get it to you for testing tomorrow. It took me a while to learn how to set up visual basic for reporting. I should have done that from the start, once I saw the error it was fixed in 5 minutes.

**Comment by Ioan-Popovici** [comment]
Released v2.3~beta [SW SQL Server Products](https://github.com/SCCM-Zone/sccm-zone.github.io/tree/master/Reporting/Software/SW%20SQL%20Server%20Products)

Please let me know if you have any problems with it.

**Comment by Ioan-Popovici** [comment]
I forgot, you’ll also need the [PivotWithDynamicColumns](https://github.com/SCCM-Zone/sccm-zone.github.io/blob/master/Reporting/User_Defined_Functions/usp_PivotWithDynamicColumns.sql) stored procedure!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by matthewgnslue** [comment]
Hi @Ioan-Popovici 
I’m testing the SW SQL Server Products.sql script.  I’m having an issue with the population of the temp table (#SQLProducts) (done with the dbo.usp_PivotWithDynamicColumns stored procedure).
The names of the VIEWS don’t match the views that I have present.  Please see the below screenshot:

![image](https://user-images.githubusercontent.com/48794619/60504496-a4d7ce00-9cb9-11e9-9479-3d12e1ccd101.png)


Is this an issue with the previously implemented version?  Which VIEWS should I use?  I assumed it was the dbo.v_GS_CUSTOM_SQL_20nn_Property_2_00 views.  However I get different results.  For example the 2017 and 2016 give different columns:

![image](https://user-images.githubusercontent.com/48794619/60504567-bde07f00-9cb9-11e9-9929-bb6324684b71.png)

Thanks.
M

**Comment by Ioan-Popovici** [comment]
The previous version is not compatible with the new one. The new one gathers far less info and is in a different format.
You need to update everything including the definitions. Do not change anything in the stored procedure.

I don’t remember if I changed the definitions to a new naming schema so they can coexist with the old definitions. That might be something I would need to address.

The correct views are v_GS_EXT.
As far as I can see you haven’t imported the new definitions.(except the product id)

 The code for the definitions should look like this:

````sql
[Dynamic, Provider("MS_VIEW_INSTANCE_PROVIDER"),
SMS_Report(true),
SMS_Group_Name("SQL 2017 Property"),
SMS_Class_ID("EXT|SQL_2017_PROPERTY|1.0")]
 
Class SQL_2017_Property: SMS_Class_Template
{
    [SMS_Report(true), key]    uint32 PropertyIndex;
    [SMS_Report(true), key]    string PropertyName;
    [SMS_Report(true)     ]    uint32 PropertyNumValue;
    [SMS_Report(true)     ]    string PropertyStrValue;
    [SMS_Report(true), key]    string ServiceName;
    [SMS_Report(true), key]    uint32 SqlServiceType;
````
Note the table name in SMS_Class_ID

These are the steps you must follow as in the blog post:

- Copy/Paste the content pf the EXT files between the extension headers in configuration.mof
- Compile the configuration.mof using mofcomp
- Copy/Paste/Run the stored procesure in SSMS (SQL Server Management Studio)
- Import the new DEF files in SCCM Hardware Inventory in Client Settings. You will need to overwrite the old definitions if the naming schema is the same.
Note if you don’t have both versions selected the old gathered dataset will be DELETED! So you will have to wait for the data to be gathered again.
- Import the new report
- Change the datasource
- Force/Schedule a client policy and HWI refresh.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Ioan-Popovici** [comment]
Hi,

There is an issue brought to my attention by @matthewgnslue .
My recommendation is to remove the old extension and replace it with the new one because the old data is not compatible with the new report. So if you don’t need the the old report and can wait for the old data to be recollected please delete the old definitions.
You can then replace the extension in configuration.mof and add a statement like the one below, for each old sql class you need to delete.
Replace the “<OLD_SQL_CLASS>” with the old extension class names.

#Pragma DeleteClass("<OLD_SQL_CLASS>", NOFAIL)

(I will update the extension with these statements once I get to a computer)

Cheers,
Ioan


<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Ioan-Popovici** [comment]
# Old Extension Cleanup Code

````sql
//----------| Old SQL extension cleanup |---------

//-- !! UNCOMMENT TO ACTIVATE !! (Remove '//' from '//#Pragma')
//#Pragma DeleteClass("SQL_2017", NOFAIL)
//#Pragma DeleteClass("SQL_2016", NOFAIL)
//#Pragma DeleteClass("SQL_2014", NOFAIL)
//#Pragma DeleteClass("SQL_2012", NOFAIL)
//#Pragma DeleteClass("SQL_2008", NOFAIL)
//#Pragma DeleteClass("SQL_2000_And_2005", NOFAIL)
````

[HWI EXT SQL Property.mof](https://github.com/SCCM-Zone/sccm-zone.github.io/blob/master/Reporting/Software/SW%20SQL%20Server%20Products/HWI%20EXT%20SQL%20Property.mof)

# Updated installation procedure

* Copy/Paste the content of the EXT files between the extension headers in configuration.mof. Overwrite the old sql extension if present or add the new extension afterwards.
**You can do it both ways, let me explain why:**
The configuration.mof file is complied on the client side and creates the extension classes. This means that machines that have already created the classes are not affected by the removal of the old extension. On the other hand, new machines added to SCCM will not create the old extension classes if the old extension does not exist in configuration.mof.
* Uncomment extension cleanup code if you want to remove the old extension classes from the clients.

**NOTE**: If you uncomment the extension cleanup code, the old extension classes will be deleted on the client side and no further data will be collected.
* Compile the configuration.mof using mofcomp
* Import the new DEF files in SCCM Hardware Inventory in Client Settings.

**NOTE:** **LEAVE THE OLD DEFINITIONS SELECTED** if you still want to use the old report, otherwise the tables will be removed from the CM DB.
![Screen Shot 2019-07-04 at 18 21 24](https://user-images.githubusercontent.com/17069425/60678244-84ac3880-9e8c-11e9-8c19-aff686c0d129.png)
You can use the old and new definitions side by side as they are using different WMI class names.
If you think you don't need the old report anymore it's safe to DELETE the old definitions.
* Copy/Paste/Run the [usp_PivotWithDynamicColumns](https://github.com/SCCM-Zone/sccm-zone.github.io/blob/master/Reporting/User_Defined_Functions/usp_PivotWithDynamicColumns.sql) stored procedure in SSMS (SQL Server Management Studio)
* Import the new report
* Change the datasource
* Force/Schedule a client policy and HWI refresh.

If you have any more questions please let me know, I will do my best to answer them.
Blog post will be updated tomorrow.

Cheers,
Ioan

**Comment by matthewgnslue** [comment]
Hi Ioan,

I've doing some testing.  I've run into an issue where executing the dbo.usp_PivotWithDynamicColumns stored procedure with the @TableName = N'dbo.v_GS_EXT_SQL_2014_Property0' does not produce the results the #SQLProducts is expecting.  Please see the below screenshot:

![image](https://user-images.githubusercontent.com/48794619/60975985-c3c30980-a324-11e9-932f-944b2ef56623.png)

**Comment by Ioan-Popovici** [comment]
I'll have a look, probably that parameter does not exist in 2014. I need change the parameter type and add a N/A if the value does not exist so the insert does nor crash and burn.

Can send me an output for the dbo.v_GS_EXT_SQL_2014_Property0 ? 1-2 machines should suffice.

I'll get on it tomorrow.

Thanks a lot for your help
Ioan

**Comment by matthewgnslue** [comment]
Thanks Ioan,

[dbo.v_GS_EXT_SQL_2014_Property0_Sample.zip](https://github.com/SCCM-Zone/sccm-zone.github.io/files/3378137/dbo.v_GS_EXT_SQL_2014_Property0_Sample.zip)

Please see the attached sample (it's a CSV file, but zipped as CSV isn't supported by GitHub).

**Comment by Ioan-Popovici** [comment]
Hmmm, no parms missing, just out of order so the insert does not know what to to with it... weird, i’ll have a closer look tomorrow.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Ioan-Popovici** [comment]
Aaah... properties are not in the same order and since the pivot is dynamic, it outputs the result in what order it found the property names. A named insert should fix it.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Ioan-Popovici** [comment]
I managed to test the query on a much bigger dataset than my current lab could offer, found the problem and also a few other bugs. (It will be in the changelog)

I used the larger dataset to validate some of the results and it seems that as far as I can tell the query returns accurate data. The mof optimization to gather only sql service data (ServiceType = 1) works better than expected, discarding a lot of junk that does not need to be filtered in the report now.

Since the extension also detects sql express and WID instances, I will add the option to filter out multiple versions in the report.

Will push an update for the query and report as soon as possible. (Most likely tomorrow)


<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Ioan-Popovici** [comment]
Hi @matthewgnslue,

I've revised the stored procedure and sql query to fix the issues below.

# Released v2.4~beta SW SQL Server Products

## Changes

* Added static column pivot to have predictable column order output
* Fixed x64 detection
* Added new EditionGroup definitions
* Default ’N/A’ if NULL Edition
* Default NULL for all CASE statements
* Removed not used columns

Please have a look if this fixes the issues.
I will test on Monday on a production environment and also update the report if everything is ok.

Cheers Ioan

**Comment by Ioan-Popovici** [comment]
# [Released v2.5 SW SQL Server Products](https://github.com/SCCM-Zone/sccm-zone.github.io/commit/10fb5e15bef5ed54386b2272bd4d73683df82be5#diff-2bf3e497db780df4efc6d9f4fb09f3ad)

## Changelist

* Added edition group filter with multiple selection
* Added CEIP Reporting (Customer Experience Improvement)
* Removed unused fields
* Fixed SKUName
* Updated formatting

Please let me know if this fixes the issue for you so I can close the issue

![Screen Shot 2019-07-15 at 15 16 55](https://user-images.githubusercontent.com/17069425/61216382-e040b680-a715-11e9-949b-b5c13047e2fd.png)

**Comment by matthewgnslue** [comment]
Hi @Ioan-Popovici ,

I can confirm that this is working for me.  I'll be doing some deeper dives into the data in the coming days.

Many thanks

**Comment by Ioan-Popovici** [comment]
Hi @matthewgnslue,

That’s great please let me know if you find any inconsistencies :)

Thanks 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Ioan-Popovici** [comment]
# Released v2.6.0 SW SQL Server Products

### Changes

* Added GetSQLProductKey.vb to the project folder
* Added the ability to hide the product key
* Fixed a typo

Closing issue
